src/producer/transactionManager.js

```diff
@@ -1,11 +1,41 @@
+const { KafkaJSNonRetriableError } = require('../errors')
+const COORDINATOR_TYPES = require('../protocol/coordinatorTypes')
+const { EventEmitter } = require('events')
+
 const NO_PRODUCER_ID = -1
 const SEQUENCE_START = 0
 const INT_32_MAX_VALUE = Math.pow(2, 32)
 
+const STATES = {
+  UNINITIALIZED: 'UNINITIALIZED',
+  READY: 'READY',
+  TRANSACTING: 'TRANSACTING',
+  COMMITTING: 'COMMITTING',
+  ABORTING: 'COMMITTING',
+}
+
+const VALID_TRANSITIONS = {
+  [STATES.UNINITIALIZED]: [STATES.READY],
+  [STATES.READY]: [STATES.READY, STATES.TRANSACTING],
+  [STATES.TRANSACTING]: [STATES.COMMITTING, STATES.ABORTING],
+  [STATES.COMMITTING]: [STATES.READY],
+  [STATES.ABORTING]: [STATES.READY],
+}
+
 /**
  * Manage behavior for an idempotent producer and transactions.
  */
-module.exports = ({ logger, cluster, transactionTimeout = 60000 }) => {
+module.exports = ({
+  logger,
+  cluster,
+  transactionTimeout = 60000,
+  transactional,
+  transactionalId,
+}) => {
+  if (transactional && !transactionalId) {
```

Comment:
I have many questions on how to pick a good `transactionalId`; I'm not saying we should set this by default but it isn't the most trivial value ever, we should think about making this easier later.

Some reference:
https://www.confluent.io/blog/transactions-apache-kafka/

> The key to fencing out zombies properly is to ensure that the input topics and partitions in the read-process-write cycle is always the same for a given transactional.id. If this isn’t true, then it is possible for some messages to leak through the fencing provided by transactions.